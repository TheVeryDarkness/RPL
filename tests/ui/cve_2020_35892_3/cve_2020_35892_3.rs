extern crate libc;

use std::ops::Index;
use std::ptr;

pub struct Slab<T> {
    capacity: usize,
    len: usize,
    mem: *mut T,
}

impl<T> Drop for Slab<T> {
    fn drop(&mut self) {
        for x in 0..self.len {
            unsafe {
                let elem_ptr = self.mem.offset(x as isize);
                ptr::drop_in_place(elem_ptr);
                std::hint::black_box(elem_ptr);
            }
        }
        unsafe { libc::free(self.mem as *mut _ as *mut libc::c_void) };
    }
}

impl<T> Index<usize> for Slab<T> {
    type Output = T;
    fn index(&self, index: usize) -> &Self::Output {
        unsafe { &(*(self.mem.offset(index as isize))) }
        //~^ERROR: it is an undefined behavior to offset a pointer using an unchecked integer
        //~|HELP:  check whether it's in bound before offsetting
        //~|HELP:  to override `-D warnings` add `#[allow(rpl::unchecked_pointer_offset)]`
    }
}

impl<T> Slab<T> {
    #[inline]
    pub fn remove(&mut self, offset: usize) -> T {
        //~^ ERROR: it usually isn't necessary to apply #[inline] to generic functions
        //~| HELP: See https://matklad.github.io/2021/07/09/inline-in-rust.html and https://rustc-dev-guide.rust-lang.org/backend/monomorph.html
        //~| HELP: to override `-D warnings` add `#[allow(rpl::generic_function_marked_inline)]`
        assert!(offset < self.len, "Offset out of bounds");

        let elem: T;
        let last_elem: T;
        let elem_ptr: *mut T;
        let last_elem_ptr: *mut T;

        unsafe {
            elem_ptr = self.mem.offset(offset as isize);
            last_elem_ptr = self.mem.offset(self.len as isize);
            //~^ HELP: this is because `self.len` exceeds the container's length by one
            //~| HELP: did you mean this

            elem = ptr::read(elem_ptr);
            last_elem = ptr::read(last_elem_ptr);
            //~^ ERROR: pointer out of bound

            ptr::write(elem_ptr, last_elem);
        }

        self.len -= 1;
        return elem;
    }
}
