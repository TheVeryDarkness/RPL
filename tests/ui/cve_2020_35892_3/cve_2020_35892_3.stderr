error: it may be an undefined behavior to offset a pointer using a passed-in integer
  --> tests/ui/cve_2020_35892_3/cve_2020_35892_3.rs:28:31
   |
LL |     fn index(&self, index: usize) -> &Self::Output {
   |                     ----- length passed in here
LL |         unsafe { &(*(self.mem.offset(index as isize))) }
   |                      -------- ^^^^^^^^^^^^^^^^^^^^^^ offset here
   |                      |
   |                      pointer used here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: `-D rpl::unchecked-pointer-offset` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::unchecked_pointer_offset)]`

error: it may be an undefined behavior to offset a pointer using a passed-in integer
  --> tests/ui/cve_2020_35892_3/cve_2020_35892_3.rs:49:33
   |
LL |     pub fn remove(&mut self, offset: usize) -> T {
   |                              ------ length passed in here
...
LL |             elem_ptr = self.mem.offset(offset as isize);
   |                        -------- ^^^^^^^^^^^^^^^^^^^^^^^ offset here
   |                        |
   |                        pointer used here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

error: pointer out of bound
  --> tests/ui/cve_2020_35892_3/cve_2020_35892_3.rs:55:25
   |
LL |             last_elem_ptr = self.mem.offset(self.len as isize);
   |                                      -------------------------
   |                                      |      |
   |                                      |      help: did you mean this: `(self.len - 1)`
   |                                      pointer created here
...
LL |             last_elem = ptr::read(last_elem_ptr);
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^ pointer read here
   |
help: this is because `self.len` exceeds the container's length by one
  --> tests/ui/cve_2020_35892_3/cve_2020_35892_3.rs:50:45
   |
LL |             last_elem_ptr = self.mem.offset(self.len as isize);
   |                                             ^^^^^^^^
   = note: `#[deny(rpl::offset_by_one)]` on by default

error: it usually isn't necessary to apply #[inline] to generic functions
  --> tests/ui/cve_2020_35892_3/cve_2020_35892_3.rs:37:5
   |
LL |       #[inline]
   |       --------- `#[inline]` here
LL | /     pub fn remove(&mut self, offset: usize) -> T {
...  |
LL | |         return elem;
LL | |     }
   | |_____^ `#[inline]` applied here
   |
   = help: See https://matklad.github.io/2021/07/09/inline-in-rust.html and https://rustc-dev-guide.rust-lang.org/backend/monomorph.html
   = note: generic functions are always `#[inline]` (monomorphization)
   = note: `-D rpl::generic-function-marked-inline` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::generic_function_marked_inline)]`

error: aborting due to 4 previous errors

